### ISSUE TITLE
staticcheck: Warn about unhandled cases when swtiching on enum value

### ISSUE THREAD

**Issue by emdash** [dominikh/go-tools#239]
Basically, add functionality similar to gcc's -Wswitch and -Wswitch-enum flags.

**Comment by dominikh** [comment]
I can see the appeal and why people would want this _sometimes_, but I think it'd be too noisy in general. 

I've encountered plenty of situations where code paths only cared about subsets of an enum, and doing nothing for unchecked enums was the right thing to do. With `-Wswitch`, adding a `default` branch would disable the check. However, I don't think that empty default branches are a good idea. They cause confusion and are prone to being removed during clean-ups. With `-Wswitch-enum`, there wouldn't be any way of allowing partial switches.

Staticcheck has very high standards when it comes to not causing false positives, in that we don't want to cause any in normal, acceptable code. Switches that only cover a subset of an enum, without a default branch, are acceptable.

Until we can figure out good heuristics for this check that get rid of most false positives, I don't see it being added.

**Comment by dominikh** [comment]
Another issue is defining what an `enum` is in Go. Go itself doesn't have enums, only constants. Conventionally, it is often assumed that this construct using iota is equivalent to enums:

```
const (
  A T = iota
  B
  C
)
```

However, if we only consider this pattern, then we can't ever have non-integer enums. String enums, for better or worse, are however a thing. At least in Go, which lacks real enums.

**Comment by dominikh** [comment]
https://www.viva64.com/en/w/v719/ mentions some useful heuristics we could use for reducing the amount of false positives.

**Comment by dominikh** [comment]
So, when I write exhaustive switches, they always have a `default` branch with a `panic` in them, so that even though the compiler won't catch my mistake, at least it'll blow up at runtime. What if we used this to detect exhaustive switches? The problem would be discerning exhaustive switches from switches that panic in their default branch for some other reason (e.g. because only a subset of all possible values is permitted). Maybe we can look for specific words in the panic message, such as `unreachable` or `exhaustive`?

**Comment by ainar-g** [comment]
<p>@dominikh&#x2003;Your original edit mentioned
<code>panic("unreachable")</code>.&#x2003;I&#xA0;think that's the&#xA0;way
to&#xA0;go, because&#xA0;I personally have never&#xA0;seen
<code>panic("exhaustive")</code>
in&#xA0;the&#xA0;wild.&#x2003;In&#xA0;the&#xA0;future we could generalise
the&#xA0;check to&#xA0;mark all reachable
<code>panic("unreachable")</code>s.</p>

**Comment by dominikh** [comment]
I agree that nobody uses the word `exhaustive` in their panic message :( Unfortunately, `unreachable` is too broad. The default branch can be unreachable even if the switch isn't exhaustive. Consider this:

```
func fn(x uint8) bool {
	if x > C {
		return false
	}
	switch x {
	case A:
	case B:
	case C:
	default:
		panic("unreachable")
	}
}
```

What we could do is give up on automatic detection, do adopt a magic comment to mark exhaustive switches, and provide a quickfix on switches that have a panicking default branch, which adds the magic comment. That way, the magic comment might be more discoverable.

**Comment by ainar-g** [comment]
<p>A&#xA0;directive like&#xA0;<code>//lint:exhaustive</code> was my first
thought as&#xA0;well, but&#xA0;so&#xA0;far <code>staticcheck</code> preferred
to&#xA0;rely almost&#xA0;exclusively on&#xA0;heuristics.&#x2003;If&#xA0;you're
ready to&#xA0;introduce a&#xA0;directive, I'm all for&#xA0;it!</p>
